File: app/src/main/java/com/aplicacion/mypet/activities/sesion/RegistroActivity.java
package com.aplicacion.mypet.activities.sesion;

import android.os.Bundle;
import android.view.View;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;

import com.aplicacion.mypet.R;
import com.google.android.gms.tasks.OnCompleteListener;
import com.google.android.gms.tasks.Task;
import com.google.android.material.textfield.TextInputEditText;
import com.google.firebase.auth.AuthResult;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.firestore.FirebaseFirestore;

import java.util.HashMap;
import java.util.Map;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class RegistroActivity extends AppCompatActivity {
    private TextInputEditText textNombre;
    private TextInputEditText textEmail;
    private TextInputEditText textPassword;
    private TextInputEditText textConfirmPassword;
    private FirebaseAuth auth;
    private FirebaseFirestore firestore;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_registro);

        textNombre = findViewById(R.id.campo_nombre);
        textEmail = findViewById(R.id.campo_email);
        textPassword = findViewById(R.id.campo_password);
        textConfirmPassword = findViewById(R.id.campo_confirmar_password);

        auth=FirebaseAuth.getInstance();
        firestore=FirebaseFirestore.getInstance();
    }

    public void irAtras(View view) {
        finish();
    }

    public void registrarUsuario(View view) {
        String nombreUsuario = textNombre.getText().toString();
        String email = textEmail.getText().toString();
        String password = textPassword.getText().toString();
        String confirmPassword = textConfirmPassword.getText().toString();

        if(!nombreUsuario.isEmpty()
                && !email.isEmpty()
                && !password.isEmpty()
                && !confirmPassword.isEmpty()) {
            if (isEmailValid(email)) {
                if (password.equals(confirmPassword)) {
                    if (password.length()>=6){
                        createUser(nombreUsuario,email,password);
                    } else {
                    Toast.makeText(this, getString(R.string.password_corta), Toast.LENGTH_LONG).show();
                    }
                } else {
                    Toast.makeText(this, getString(R.string.password_diferente), Toast.LENGTH_LONG).show();
                }
            } else {
                Toast.makeText(this, getString(R.string.email_no_valido), Toast.LENGTH_LONG).show();
            }
        } else {
            Toast.makeText(this, getString(R.string.campos_vacios), Toast.LENGTH_LONG).show();
        }
    }

    private void createUser(final String nombreUsuario,final String email, String password) {
        auth.createUserWithEmailAndPassword(email,password).addOnCompleteListener(new OnCompleteListener<AuthResult>() {
            @Override
            public void onComplete(@NonNull Task<AuthResult> task) {
                if(task.isSuccessful()){
                    addDataUser(email,nombreUsuario);
                }else{
                    Toast.makeText(RegistroActivity.this, getString(R.string.register_incomplete), Toast.LENGTH_LONG).show();
                }
            }
        });
    }

    private void addDataUser(String email,String nombreUsuario){
        String id = auth.getCurrentUser().getUid();
        Map<String,Object> datosUsuario = new HashMap<>();
        datosUsuario.put("email",email);
        datosUsuario.put("nombreUsuario",nombreUsuario);
        firestore.collection("Users").document(id).set(datosUsuario).addOnCompleteListener(new OnCompleteListener<Void>() {
            @Override
            public void onComplete(@NonNull Task<Void> task) {
                if(task.isSuccessful()) {
                    Toast.makeText(RegistroActivity.this, getString(R.string.register_complete), Toast.LENGTH_LONG).show();
                    finish();
                }
                else
                    Toast.makeText(RegistroActivity.this, getString(R.string.register_incomplete), Toast.LENGTH_LONG).show();
            }
        });
    }

    private boolean isEmailValid(String email) {
        String expression = "^[\\w\\.-]+@([\\w\\-]+\\.)+[A-Z]{2,4}$";
        Pattern pattern = Pattern.compile(expression, Pattern.CASE_INSENSITIVE);
        Matcher matcher = pattern.matcher(email);
        return matcher.matches();
    }
}

File: app/src/main/java/com/aplicacion/mypet/activities/sesion/RegistroActivity.kt
package com.aplicacion.mypet.activities.sesion

import android.app.AlertDialog
import android.os.Bundle
import android.view.View
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import com.aplicacion.mypet.R
import com.aplicacion.mypet.models.User
import com.aplicacion.mypet.providers.AuthProvider
import com.aplicacion.mypet.providers.UserProvider
import com.google.android.material.textfield.TextInputEditText
import dmax.dialog.SpotsDialog
import java.util.regex.Pattern

class RegistroActivity : AppCompatActivity() {
    private lateinit var mTextNombre: TextInputEditText
    private lateinit var mTextEmail: TextInputEditText
    private lateinit var mTextPassword: TextInputEditText
    private lateinit var mTextConfirmPassword: TextInputEditText
    private lateinit var mAuth: AuthProvider
    private lateinit var mUserProvider: UserProvider
    private lateinit var mDialog: AlertDialog

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_registro)

        mTextNombre = findViewById(R.id.campo_nombre)
        mTextEmail = findViewById(R.id.campo_email)
        mTextPassword = findViewById(R.id.campo_password)
        mTextConfirmPassword = findViewById(R.id.campo_confirmar_password)

        mAuth = AuthProvider()
        mUserProvider = UserProvider()

        mDialog = SpotsDialog.Builder()
                .setContext(this)
                .setMessage(R.string.cargando_registro)
                .setCancelable(false).build()

    }

    fun irAtras(view: View?) {
        finish()
    }

    fun registrarUsuario(view: View?) {
        val nombreUsuario = mTextNombre.text.toString()
        val email = mTextEmail.text.toString()
        val password = mTextPassword.text.toString()
        val confirmPassword = mTextConfirmPassword.text.toString()
        if (nombreUsuario.isNotEmpty()
                && email.isNotEmpty()
                && password.isNotEmpty()
                && confirmPassword.isNotEmpty()) {
            if (isEmailValid(email)) {
                if (password == confirmPassword) {
                    if (password.length >= 6) {
                        createUser(nombreUsuario, email, password)
                    } else {
                        Toast.makeText(this, getString(R.string.password_corta), Toast.LENGTH_LONG).show()
                    }
                } else {
                    Toast.makeText(this, getString(R.string.password_diferente), Toast.LENGTH_LONG).show()
                }
            } else {
                Toast.makeText(this, getString(R.string.email_no_valido), Toast.LENGTH_LONG).show()
            }
        } else {
            Toast.makeText(this, getString(R.string.campos_vacios), Toast.LENGTH_LONG).show()
        }
    }

    private fun createUser(nombreUsuario: String, email: String, password: String) {
        mDialog.show()
        mAuth.registerUser(email, password).addOnCompleteListener { task ->
            if (task.isSuccessful) {
                addDataUser(email, nombreUsuario)
            } else {
                Toast.makeText(this@RegistroActivity, getString(R.string.register_incomplete), Toast.LENGTH_LONG).show()
            }
        }
    }

    private fun addDataUser(email: String, nombreUsuario: String) {
        val id = mAuth.uid
        val user = User()
        user.id = id
        user.email = email
        user.username = nombreUsuario
        mUserProvider.create(user).addOnCompleteListener { task ->
            if (task.isSuccessful) {
                mDialog.dismiss()
                Toast.makeText(this@RegistroActivity, getString(R.string.register_complete), Toast.LENGTH_LONG).show()
                finish()
            } else Toast.makeText(this@RegistroActivity, getString(R.string.register_incomplete), Toast.LENGTH_LONG).show()
        }
    }

    private fun isEmailValid(email: String): Boolean {
        val expression = "^[\\w.-]+@([\\w\\-]+\\.)+[A-Z]{2,4}$"
        val pattern = Pattern.compile(expression, Pattern.CASE_INSENSITIVE)
        val matcher = pattern.matcher(email)
        return matcher.matches()
    }

}